Merging HIG-16-006

ajgilbert opened the pull request:
Preparing merge back into master

ajgilbert commented:
Changes outside the HIG16006 area:
-  Added a ch::contains function to Algorithms.h that works with
   initializer lists, e.g: if (ch::contains({"A","B","C", "C"}) (8fdf82ef79f646167c20ebdc0740d9d803c07f8b)
- Fixed bug in CH evaluate methods which were not ignoring rateParam Systematic entries (1291618ebaf2639cb09b7fb792ba1a44f38eee10)
- Fixed minor bug in plotImpacts.py (c0f225c7da3f14a2e5733a5e9eeca41445162845)
- Corrected iteration through TH1 bins in plotting.py function `GetPadYMaxInRange` (71497384448445270f4a92b9bae57dba3515dc92)
- Added `--exclude` option for parameters in the Impacts combineTool.py module (61e24f40f530319b12c89892e801d8fd04601f2d)
- Added `--units` option for the POI fit result in plotImpacts.py (61e24f40f530319b12c89892e801d8fd04601f2d)
- Reworked the boundlist implementation in EnhancedCombine (58c06a3c9da129b642ce2846b87841977bc463b7, 67d4b4aee45e22f8f58fea3c80939961c37ebfc8)
- Added option to BinByBinFactory to create poisson-interval uncertainties, off by default (877362fc9a84e3a348f2cfa441951c32a103d4dc, c17035361d16b638d911eeb27683a359855a317c)
- AutoRebin: added additional mode 2 that only merges left, was needed for review question but probably not useful in general. Also updated the printout info slightly (9b8f4f54b640119bfbb2d3537fa4b47512438e52, 30dab8149519ec685327830b1ddb7a9da8fb6b7c)
- CollectGoodnessOfFit method added to combineTool.py and plotting script `plotGof.py` added (479b65ff69679293a9c1675349ecef17ddc4b909)
- Added option to BinByBinFactory to skip merging bin errors when the bin content is zero. Disabled by default so no change in normal behaviour. (6d1f78bc18874bcce7a7b1f3f9f13e29f13a766a)
- plotting.py function `fastFillTH2` for quicker filling from a TGraph2D when bin centres are aligned with graph points (ce7356addd5697201830bc4af3852bb15d6be96b)
- Plotting script for 2D NLL scans: `CombineTools/scripts/plotMultiDimFit.py` (ce7356addd5697201830bc4af3852bb15d6be96b, 89f557a3f299f16ca78b1d5e8b9b7cdb6fad8707, bb7fb0896cbabd973e91f9b462ca46579c73082d, b5d89c262f7bcf8c507a1a51c779708d18d1d01b)
- BuildRooMorphing: save norm TGraph stat uncertainties in the debug output (67d4b4aee45e22f8f58fea3c80939961c37ebfc8)
- CombineHarvester: added GetRateCovariance and GetRateCorrelation methods (3e8120df28b1b355dc4ad80654bc1ddb836aeeee)
- PostFitShapesFromWorkspace updates:
  * option for adding covariance matrix of process norms to output (3e8120df28b1b355dc4ad80654bc1ddb836aeeee)
  * prints individual process yields (3e8120df28b1b355dc4ad80654bc1ddb836aeeee)
  * added option to specify a non-default dataset (e542bf23c4c5c7187305b9f0114f2397ae1ea9aa)
- Update for crab submission in combineTool.py (cf767d70bbed4fd8a5e7e913f10f1ec4f142648a, 036fd1867f94f1776f4a116f1954fe879e8de966)
- HybridNewGrid updates:
  * Extra option to remove specific points from the grid (e5ce142819f98c88c9b489bf058cd8ea4292eb51, 008bf0ea5f89c7f423c016c3fa1444aebe90cb3c)
  * Optional json status file can be used to skip re-calculating all CLs checks when none of the input files for a point have changed (e74dd38c13017ca2774ac6bdea1e342d890dbfc6, 8833a1495d0617996fa624a80c11eaf25a802460)
- Updated `CombineTools/scripts/plotLimitGrid.py` for final results (326a9aaa8c4b4c9a1190751a6666ddb1f9635310, a9c17ea5bc2720a99782ffd9a686611557346e30, a7d45bb3553d12cf973bbee26c56acda49fc15b9, c50523d51115840fac1831a59b49a1a12f21d61f, 2fd9030427dfdf226eb8be85a8f225bb3bd0992b, 215fe611ed83dc8d7b0befbf4372562811bb5e25)
- Bugfix in combineTool.py for using `--there` with `--merge` (d02a029abf769178a6ebfb1b01a5f290da202cd9)
- CombineHarvester: allow substitution of `$BINID` in systematic names when using `AddSyst` (a7ae05dc8ec0e4cb8172b8266ed37c67fa8d640d)
- combineTool.py method for collected limit toys (e8b6b5efa0dde9f544fb5ec283ac715017ef455c) *duplicated implementation with 3fa1414fc1828ba71fa11c7d9b98474ec1f0d89c*
- combineTool.py T2W method now sorts list of datacard inputs to make sure combined card is fully deterministic (aa90dec5149793c166c9288d1db0ca773391470a)


Remaining tasks before merge:

 - [ ] Check if a slowdown has been inadvertently introduced into PostFitShapesFromWorkspace (commits 3e8120df28b1b355dc4ad80654bc1ddb836aeeee, Y and Z)
 - [ ] Move `CombineTools/src/HttSystematics_MSSMRun2.cc` to HIG16006 area
 - [ ] Put `BinByBinFactory::SetPoissonErrors` in the python interface
 - [ ] If `CombineTools/scripts/plotMultiDimFit.py` has analysis-specific code in should be moved to HIG16006 for now
 - [ ] If `CombineTools/scripts/plotLimitGrid.py` has analysis-specific code in should be moved to HIG16006 for now
 - [ ] Don't want analysis-specific limit band drawing options: https://github.com/cms-analysis/CombineHarvester/pull/51/commits/e8b6b5efa0dde9f544fb5ec283ac715017ef455c#diff-6f971de980e397e2ad7c2f6fbd3c23b6 (also e229c0cdc2c77efd75e300164dd5f407bb59a68f)
 - [ ] Resolve duplicated toy limit collection implementations: 3fa1414fc1828ba71fa11c7d9b98474ec1f0d89c and e8b6b5efa0dde9f544fb5ec283ac715017ef455c

rcaspart commented:
For the second to last point (no analysis specific drawing options) I would suggest to merge #53 first. This pull request could also be further enhanced to allow to plug in a dedicated style-files to facilitate an even more analysis independent script.

ajgilbert commented:
@rcaspart last three commits adjust the GoF code a bit - I think everything should still work the same but if you get a chance to check it would be good.

ajgilbert commented:
Final changes before merge:

 - Have reverted some changes to PostFitShapesFromWorkspace that were increasing the time by 10-20%, now roughly the same as the master branch again
 - Moved plotMultiDimFit.py to HIG16006 from CombineTools for now as it only seems to be usable when r_ggH and r_bbH are the POIs (can easily be made more generic later and moved back)
